File: app/src/main/java/com/example/warehousewrangler/WarehouseViewModel.kt
package com.example.warehousewrangler

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import com.example.warehousewrangler.models.WarehouseIssue
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class WarehouseViewModel(private val repository: DynamoDBManager) : ViewModel() {

    private val _currentIssue = MutableLiveData<WarehouseIssue?>()
    val currentIssue: LiveData<WarehouseIssue?> = _currentIssue

    private val _statusMessage = MutableLiveData<String>()
    val statusMessage: LiveData<String> = _statusMessage

    private val _toastMessage = MutableLiveData<String>()
    val toastMessage: LiveData<String> = _toastMessage

    private val _userId = "USER_001" // Hardcoded for now

    fun loadNextTask() {
        _statusMessage.postValue("Status: Loading...")
        repository.fetchAssignedIssue(_userId) { issue ->
            if (issue != null) {
                if (issue.status == "ASSIGNED") {
                    issue.status = "SEARCHING"
                    repository.updateIssue(issue) {
                        // Silent update
                    }
                }
                _currentIssue.postValue(issue)
                _statusMessage.postValue("Status: Searching")
            } else {
                _currentIssue.postValue(null)
                _statusMessage.postValue("Status: No Tasks Assigned")
            }
        }
    }

    fun processScan(data: String) {
        val issue = _currentIssue.value ?: return

        if (data == issue.skuId) {
            issue.currentQty += 1
            _currentIssue.postValue(issue) // Trigger update

            if (issue.currentQty >= issue.targetQty) {
                completeIssue(issue)
            } else {
                _toastMessage.postValue("Item Scanned. ${issue.currentQty}/${issue.targetQty}")
            }
        } else {
            _toastMessage.postValue("Wrong SKU!")
        }
    }

    private fun completeIssue(issue: WarehouseIssue) {
        issue.status = "FOUND"
        _statusMessage.postValue("Status: Item Found! Updating...")

        repository.updateIssue(issue) { success ->
            if (success) {
                _toastMessage.postValue("Task Completed!")
                loadNextTask()
            } else {
                _toastMessage.postValue("Error Updating DB")
            }
        }
    }

    fun markAsMissing() {
        val issue = _currentIssue.value ?: return
        issue.status = "MISSING"
        _statusMessage.postValue("Status: Marking Missing...")

        repository.updateIssue(issue) { success ->
            if (success) {
                _toastMessage.postValue("Marked Missing")
                loadNextTask()
            } else {
                _toastMessage.postValue("Error Updating DB")
            }
        }
    }
}

class WarehouseViewModelFactory(private val repository: DynamoDBManager) : ViewModelProvider.Factory {
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(WarehouseViewModel::class.java)) {
            @Suppress("UNCHECKED_CAST")
            return WarehouseViewModel(repository) as T
        }
        throw IllegalArgumentException("Unknown ViewModel class")
    }
}

File: ios/WarehouseWranglerIOS/Sources/ViewModel/WarehouseViewModel.swift
import Foundation

extension Notification.Name {
    static let warehouseStateDidChange = Notification.Name("WarehouseViewModel.warehouseStateDidChange")
    static let warehouseToast = Notification.Name("WarehouseViewModel.warehouseToast")
}

final class WarehouseViewModel {

    private let repository: WarehouseRepository
    private let userId: String

    private(set) var currentIssue: WarehouseIssue?
    private(set) var stage: PickScanStage = .confirmLocation
    private(set) var statusMessage: String = "Status: Idle"
    private(set) var lastScan: String = "None"

    init(repository: WarehouseRepository, userId: String) {
        self.repository = repository
        self.userId = userId
    }

    func loadNextTask() {
        statusMessage = "Status: Loading..."
        notifyStateChanged()

        repository.fetchAssignedIssue(userId: userId) { [weak self] issue in
            guard let self else { return }
            DispatchQueue.main.async {
                if var issue = issue {
                    if issue.status == "ASSIGNED" {
                        issue.status = "SEARCHING"
                        self.repository.updateIssue(issue) { _ in }
                    }

                    self.currentIssue = issue
                    self.stage = .confirmLocation
                    self.statusMessage = "Status: Scan Location"
                } else {
                    self.currentIssue = nil
                    self.stage = .confirmLocation
                    self.statusMessage = "Status: No Tasks Assigned"
                }

                self.lastScan = "None"
                self.notifyStateChanged()
            }
        }
    }

    func processScan(_ data: String) {
        let scanned = data.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !scanned.isEmpty else { return }

        lastScan = scanned

        guard var issue = currentIssue else {
            notifyStateChanged()
            return
        }

        let targetLocation = issue.targetLocation.trimmingCharacters(in: .whitespacesAndNewlines)

        if stage == .confirmLocation {
            if scanned.caseInsensitiveCompare(targetLocation) == .orderedSame {
                stage = .scanItem
                statusMessage = "Status: Scan Item"
                postToast("Location confirmed: \(targetLocation)")
            } else {
                postToast("Wrong Location!")
            }
            notifyStateChanged()
            return
        }

        if scanned.caseInsensitiveCompare(targetLocation) == .orderedSame {
            postToast("Location confirmed: \(targetLocation)")
            notifyStateChanged()
            return
        }

        let matchesSku = scanned.caseInsensitiveCompare(issue.skuId) == .orderedSame ||
            scanned.caseInsensitiveCompare(issue.barcode) == .orderedSame

        if matchesSku {
            issue.currentQty += 1
            currentIssue = issue
            notifyStateChanged()

            if issue.currentQty >= issue.targetQty {
                completeIssue(issue)
            } else {
                postToast("Item scanned. \(issue.progressText)")
            }
        } else {
            postToast("Wrong SKU!")
            notifyStateChanged()
        }
    }

    func markMissing() {
        guard var issue = currentIssue else { return }
        issue.status = "MISSING"
        currentIssue = issue
        statusMessage = "Status: Marking Missing..."
        notifyStateChanged()

        repository.updateIssue(issue) { [weak self] success in
            DispatchQueue.main.async {
                if success {
                    self?.postToast("Marked Missing")
                    self?.loadNextTask()
                } else {
                    self?.postToast("Error updating task")
                }
            }
        }
    }

    private func completeIssue(_ issue: WarehouseIssue) {
        var issue = issue
        issue.status = "FOUND"
        currentIssue = issue
        statusMessage = "Status: Task Found! Updating..."
        notifyStateChanged()

        repository.updateIssue(issue) { [weak self] success in
            DispatchQueue.main.async {
                if success {
                    self?.postToast("Task Completed!")
                    self?.loadNextTask()
                } else {
                    self?.postToast("Error updating task")
                }
            }
        }
    }

    private func notifyStateChanged() {
        if Thread.isMainThread {
            NotificationCenter.default.post(name: .warehouseStateDidChange, object: self)
        } else {
            DispatchQueue.main.async {
                NotificationCenter.default.post(name: .warehouseStateDidChange, object: self)
            }
        }
    }

    private func postToast(_ message: String) {
        if Thread.isMainThread {
            NotificationCenter.default.post(
                name: .warehouseToast,
                object: self,
                userInfo: ["message": message]
            )
        } else {
            DispatchQueue.main.async {
                NotificationCenter.default.post(
                    name: .warehouseToast,
                    object: self,
                    userInfo: ["message": message]
                )
            }
        }
    }
}

